[DoctrineBridge] Add support for doctrine/persistence 4

greg0ire opened the pull request:
[v4](https://github.com/doctrine/persistence/blob/4.0.x/UPGRADE.md#upgrade-to-40) provides a guarantee that `ManagerRegistry::getManager()` returns an entity manager (as opposed to null).

| Q             | A
| ------------- | ---
| Branch?       | 6.4
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Issues        | n/a
| License       | MIT

greg0ire committed:
Add support for doctrine/persistence 4

v4 provides a guarantee that ManagerRegistry::getManager() returns an entity
manager (as opposed to null).

The tests need to be adjusted to reflect the behavior of the mocked dependency
more accurately, as it is throwing an exception in case of a null manager for
all three supported versions of the library.

greg0ire commented:
Let me know if this is the wrong branch, I'm never quite sure for this kind of case with Symfony, but in the past, I think I've been asked to contribute this kind of branch as a patch.

derrabus commented on src/Symfony/Bridge/Doctrine/Tests/ArgumentResolver/EntityValueResolverTest.php:
```diff
@@ -54,6 +54,10 @@ public function testResolveWithoutAttribute()
 
     public function testResolveWithoutManager()
     {
+        if (method_exists(ObjectManager::class, 'isUninitializedObject')) {
+            $this->markTestSkipped('This test is not applicable to doctrine/persistence 4');
+        }
```
How does Persistence 4 behave in this situation? Should we have a test for that as well?

stof replied:
@derrabus `TypeError` in the signature of the `AbstractManagerRegistry` when passing `null` as default manager name. And `InvalidArgumentException` if there is no manager with that name.

@greg0ire we cannot remove support for the case of having no entity manager. As the ManagerRegistry interface extends ConnectionRegistry in doctrine/persistence (which is a mistake to me, but won't be an easy fix) and we don't have any standalone implementation of ConnectionRegistry, DoctrineBundle will still create a manager registry even when your project enables DBAL and not ORM. And so this case exists (it won't be a case of having no manager registry at all)

stof replied:
the proper fix is to fix the mock of the ManagerRegistry. When passing `null` as argument here, `getManagerForClass` should indeed return `null` but `getManager` should throw an `InvalidArgumentException` (which is already the behavior of doctrine/persistence 2.x and 3.x btw)

greg0ire replied:
Thanks for the pointers, I'll try that 👍

stof commented on src/Symfony/Bridge/Doctrine/ArgumentResolver/EntityValueResolver.php:
```diff
@@ -50,6 +50,7 @@ public function resolve(Request $request, ArgumentMetadata $argument): array
         if (!$options->class || $options->disabled) {
             return [];
         }
+        // doctrine/persistence < 4 compat
```
This is not true. `$this->getManager` can still return `null` (see the implementation of the private method). This method call is **not** a call to the doctrine/persistence method.

greg0ire replied:
Oh right, didn't spot that!

stof reviewed (changes requested):
This PR is broken in its current state. It should update mocks to match the actual behavior of doctrine/persistence instead of skipping relevant tests.

-1 for merging it in its current state.

greg0ire commented:
@stof I've addressed your review.

Also, I forgot to mention it, but `doctrine/persistence` 4 is now indeed installed in at least [this job](https://github.com/symfony/symfony/actions/runs/12914277977/job/36013506034?pr=59575#step:5:60)

stof commented on src/Symfony/Bridge/Doctrine/Tests/ArgumentResolver/EntityValueResolverTest.php:
```diff
@@ -415,9 +415,16 @@ private function createRegistry(?ObjectManager $manager = null): ManagerRegistry
-        $registry->expects($this->any())
-            ->method('getManager')
-            ->willReturn($manager);
+        if (method_exists(ObjectManager::class, 'isUninitializedObject')) {
```
This should check `$manager === null`. We don't want to throw when we have an instance.

And I suggest applying this condition even for older versions of doctrine/persistence so that the mock reflects the actual behavior of the class. AbstractManagerRegistry was already throwing an exception in persistence 2.x (I did not check the behavior in 1.x)

greg0ire replied:
> I did not check the behavior in 1.x

It is not allowed by Symfony's `composer.json` 👍

stof replied:
I know. That's exactly why I did not check 1.x to see whether our mocks were already wrong back in time or whether they kept emulating the 1.x behavior, because distinguishing those 2 cases does not matter.

stof commented on src/Symfony/Bridge/Doctrine/Tests/Validator/Constraints/UniqueEntityValidatorTest.php:
```diff
@@ -635,6 +639,10 @@ public function testDedicatedEntityManagerNullObject()
     public function testEntityManagerNullObject()
     {
+        if (method_exists(ObjectManager::class, 'isUninitializedObject')) {
+            $this->markTestSkipped('This test is not applicable to doctrine/persistence 4');
```
Those tests are application, for the same reason that the EntityValueResolver test (and with the same issue about a bad definition of the mock)

stof commented on src/Symfony/Bridge/Doctrine/Validator/Constraints/UniqueEntityValidator.php:
```diff
@@ -71,6 +71,7 @@ public function validate(mixed $entity, Constraint $constraint)
         if ($constraint->em) {
             $em = $this->registry->getManager($constraint->em);
 
+            // doctrine/persistence < 4 compat
```
this comment is actually wrong. Older versions of `doctrine/persistence` are also throwing an exception in `getManager`

greg0ire commented on src/Symfony/Bridge/Doctrine/Validator/Constraints/UniqueEntityValidator.php:
```diff
@@ -69,10 +69,14 @@ public function validate(mixed $entity, Constraint $constraint)
-            $em = $this->registry->getManager($constraint->em);
-
-            if (!$em) {
-                throw new ConstraintDefinitionException(sprintf('Object manager "%s" does not exist.', $constraint->em));
```
I removed the handling of `null` here, since it is not really supposed to happen. Let me know if that's wrong.

stof replied:
that's totally fine, as it can only happen with a bad mock.

fabpot commented on src/Symfony/Bridge/Doctrine/Validator/Constraints/UniqueEntityValidator.php:
```diff
@@ -69,10 +69,14 @@ public function validate(mixed $entity, Constraint $constraint)
+                    sprintf('Object manager "%s" does not exist.', $constraint->em),
+                    0,
+                    $e
+                );
```
Should be on one line.

fabpot commented:
Thank you @greg0ire.
